File: apps/desktop/src/mappers/graph.ts
import type { GraphEdgeResponse, GraphEntityResponse, GraphSnapshotResponse } from '@nova/api-contracts'
import type { GraphEdge, GraphEntity, GraphViewModel } from '@/view-models'

export function mapGraphEntity(row: GraphEntityResponse): GraphEntity {
  return {
    id: row.id,
    type: row.type,
    canonicalName: row.canonical_name,
    attributes: row.attributes,
  }
}

export function mapGraphEdge(row: GraphEdgeResponse): GraphEdge {
  return {
    id: row.id,
    fromEntityId: row.from_entity_id,
    toEntityId: row.to_entity_id,
    relationType: row.relation_type,
    weight: row.weight,
    sourceMemoryId: row.source_memory_id ?? undefined,
  }
}

export function mapGraph(response: GraphSnapshotResponse): GraphViewModel {
  return {
    entities: response.entities.map(mapGraphEntity),
    edges: response.edges.map(mapGraphEdge),
  }
}

File: apps/backend/services/api/schemas/graph.py
from typing import Any, Optional

from pydantic import BaseModel, Field


class GraphEntityResponse(BaseModel):
    id: str
    type: str
    canonical_name: str
    attributes: dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[str] = None


class GraphEdgeResponse(BaseModel):
    id: str
    from_entity_id: str
    to_entity_id: str
    relation_type: str
    weight: float
    source_memory_id: Optional[str] = None
    created_at: Optional[str] = None


class GraphRelatedResponse(BaseModel):
    entity: GraphEntityResponse
    related: list[GraphEntityResponse]
    edges: list[GraphEdgeResponse] = Field(default_factory=list)


class GraphSnapshotResponse(BaseModel):
    entities: list[GraphEntityResponse] = Field(default_factory=list)
    edges: list[GraphEdgeResponse] = Field(default_factory=list)


class GraphStatsResponse(BaseModel):
    entity_count: int
    edge_count: int
    entities_by_type: dict[str, int] = Field(default_factory=dict)
    edges_by_relation: dict[str, int] = Field(default_factory=dict)
